### PR TITLE
Initial run of network layer

### PR DESCRIPTION
**Pull request by jbaxleyiii**
Initial run of the network layer to allow for remote requests. This interface will be used by the end client API to get the benefits of the caching system and slimmer requests.

### CODE REVIEW

**Comment by stubailo** [comment]
@jbaxleyiii should we be worried about the patent clause in Relay? Might it be worth avoiding taking parts from it? Maybe I'm overthinking it.

@zol looks like the CLA bot is broken, it's reporting not signed even though it is according to the website.

File: src/networkLayer.js [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+import {
+  isString,
+  isArray,
+} from 'lodash';
+
+class NetworkLayer {
+
+  constructor(uri, opts = {}) {
+    if (!uri) {
+      throw new Error('A remote enpdoint is required for a newtork layer');
```
Comment (stubailo) [inline src/networkLayer.js]: endpoint

File: src/networkLayer.js [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+    }
+
+    this._uri = uri;
+    this._opts = { ...opts };
+  }
+
+  query(requests) {
+    let clonedRequests = [];
+
+    if (!isArray(requests)) {
```
Comment (stubailo) [inline src/networkLayer.js]: Personally I'm not a fan of APIs that take an array or a single item, what do you think? If we want this to make multiple requests IMO we may as well have it always take an array.

A counter-argument could be if we are completely compatible with Relay network layers, that would be a feature. Is this 100% compatible with a network layer written for Relay?

**Comment by jbaxleyiii** [comment]
@stubailo I'll take a look at the patent. I can make it more generic as well to avoid any complications

File: src/networkLayer.js (hunk 2 above)
Comment (jbaxleyiii) [inline src/networkLayer.js]: Right now relay only accepts an array, so I'll scrap the single item check so its compatible and a single typed argument

**Comment by stubailo** [comment]
@jbaxleyiii I dunno if the patent thing is a big deal. Maybe I should consult with someone who knows about this stuff to see how much we should care. At the end of the day everyone uses React and it has a similar patent grant.

**Comment by jbaxleyiii** [comment]
@stubailo that sounds good. Keeping it compatible as a relay network layer would be cool, but I don't know how much value it would add since it this layer will probably stay bare bones. If we don't worry about it, we can change the `request` object to better suit apollo's needs and add in additional upgrades as needed without concern of breaking relay compatibility

**Comment by stubailo** [comment]
Yeah makes sense to me. Ensuring compatibility with the Relay client doesn't sound like a high priority right now because I haven't seen that many libraries written for it yet.

**Comment by stubailo** [comment]
@jbaxleyiii I think the caching wouldn't be integrated into the request/network layer right?

I'm thinking:

```
client
- cache
- network layer
```

Where the cache and network layer are swappable, and the client uses both and mediates the communication. The network layer just does queries when the client wrapper deems it necessary?

**Comment by jbaxleyiii** [comment]
@stubailo yep! That was what I was going to work on next. That also potentially allows us a few other options:

```
client
- cache
- network layer 1 (third party GQL endpoint)
- network layer 2 (first party GQL endpoint)
- application state endpoing // this one is probably off spec, but for uniformity I would love to setup my application state to be query-able / modified via query and mutations
```

**Comment by stubailo** [comment]
Yep, sounds good to me. One thing I'm looking into now is the ability to, given a cache and a query, figure out the set of queries needed to fetch the data that isn't in the cache yet. Sounds like that will be useful to avoid fetching data we already have.

> application state endpoint

Definitely interested in what this looks like!

**Comment by jbaxleyiii** [comment]
@stubailo I think the client api is a larger build out than just the network layer. I'm going to refactor this slightly tonight to make it a little cleaner / remove some the relay code. Then we should be good to merge it through and start proper on the client api

**Comment by stubailo** [comment]
Yep, I completely agree. I don't think this PR should include anything but a network layer implementation. Also we should agree on a design for the client before we start implementing one, since it's kind of what brings it all together.

**Comment by jbaxleyiii** [comment]
@stubailo this should be good to go code wise. At some point we should run a local server for testing instead of using the swapi graphql.

**Pull request merged**